### PR TITLE
wip: Adds basic scheduler

### PR DESCRIPTION
**Pull request by przemyslaw-przylucki**
Quite a bit more to do here but would appreciate any direction feedback @brendt

### CODE REVIEW

File: src/ConsoleCommand.php
Comment (przemyslaw-przylucki) [inline src/ConsoleCommand.php:22]: for convenience this could accept either crondefinition, a cron string or dateinterval and then normalize it to CronDefinition

Comment (brendt) [reply src/ConsoleCommand.php:22]: What I envisioned is only one way to avoid confusion. You can read some more about my vision on this topic here: https://stitcher.io/blog/opinion-driven-design

Comment (brendt) [inline src/ConsoleCommand.php:22]: I think `schedule` is a better word than `cron`, but I'm not entirely against `cron` either. wdyt?

File: src/Scheduler/CronDefinition.php
Comment (brendt) [inline src/Scheduler/CronDefinition.php:9]: I don't really like this name. Either it should be `Schedule` or `Cron`, my preference is the first.

File: src/Scheduler/CronDefinition.php
Comment (brendt) [inline src/Scheduler/CronDefinition.php:11]: Could you explain these four option? In my mind, we only need an interval; the process will always run in the background, and I don't think output should be configurable on this level (maybe on config level it should)

File: src/Scheduler/GenericScheduler.php
Comment (brendt) [inline src/Scheduler/GenericScheduler.php:38]: I don't like this approach of setting instance variables in one function that influence the behaviour of other functions. I think the better approach is to pass these variables explicitly to `getCommandsToRun`

File: src/Scheduler/GenericScheduler.php
Comment (brendt) [inline src/Scheduler/GenericScheduler.php:57]: I understand you went with a stateless approach. Basically checking if the interval matches the current time. 

I want to pitch another stateful approach, that has a couple of benefits. Would love your input.

Let's say we keep track of when a command has last run. That means that determining whether a command should run or not is as simple as checking whether the interval is smaller than the difference between now and the last run.

The benefit here is that if the scheduler fails to run for an amount of time, it'll be able to start up again and finish tasks that should have been run while it was down. 

The downside is that you loose fine-grained control over _when_ a command should run. If you want it to run exactly every 5th minute after the hour, you're out of luck. 

We could of course provide _both_ options, although I would start with the stateful approach, since it significantly reduces complexity in determining whether a scheduled command should run or not. 

wdyt?

File: src/Scheduler/CronDefinition.php
Comment (przemyslaw-przylucki) [reply src/Scheduler/CronDefinition.php:11]: Yeah, I kinda was wondering about the approach we'd want to take here.

Generally speaking if we're going with "the framework that gets out of your way" narrative, we'd probably want to drop this class altogether and just use the DateInterval directly in command method's declaration, the problem I see is that commands are very different from each other and the way people use them will vary.

Some people will use them to just grab some DB records and dispatch a bunch of jobs, some people will use them to modify a status of an entity, where they have one command per status i.e InitializePaymentsCommand, FetchPaymentStatusCommand, SendPaymentConfirmationCommand, and they'd want to have them run in order - not really a fan of that solution myself, but I've seen that in several projects.

What I didn't think about is that wouldn't even be a thing in tempest because we cannot specify order with discovery, so that's not really valid 😅

For other things though - like being able to change an output of specific command to another file, or doing some other customizations may come in handy in the future.

So I'd suggest keeping this class, so we can extend the scheduler without breaking the api, seeing how things like output, or whether or not job could run cross-servers (if that should even be a concern in Tempest, this really depends on the bigger picture) could be important to modify on per-command basis.

Idk how you feel about this, and this def has a hacky feeling to it, but maybe it'd be nice to separate the concept of running something in intervals as a first party citizien.

This way, if we have some simple prunner in our service, we could simply do

```php

#[Interval(minute: 1)]
public function prune(): void
{
   DB::table('messages')->where('created_at', '<=', now())->delete();
}

``` 

This has it's ups and downs and creates more magic but it feels like a nice utility

File: src/Scheduler/GenericScheduler.php
Comment (przemyslaw-przylucki) [reply src/Scheduler/GenericScheduler.php:57]: That makes complete sense!

How would you want to handle the saving of last command run? Do we want to create a lockfile, handle that in the DB or through cache?

File: src/Scheduler/CronDefinition.php
Comment (brendt) [reply src/Scheduler/CronDefinition.php:11]: > and they'd want to have them run in order

This shouldn't be a concern of the scheduler; there are other patterns more suited to handling those. One that comes to mind are [sagas](https://learn.microsoft.com/en-us/azure/architecture/reference-architectures/saga/saga) — which are definitely out of scope for the projects Tempest is targeting.

> may come in handy in the future.

It could be, but right now we definitely not going down that route, so it's adding quite a lot of boilerplate for something that might happen. 

> Idk how you feel about this, and this def has a hacky feeling to it, but maybe it'd be nice to separate the concept of running something in intervals as a first party citizien.

Why does it have a hacky feeling? I think your last code block is the way to go, although I'd change the name to `Schedule` for clarity

File: src/Scheduler/GenericScheduler.php
Comment (brendt) [reply src/Scheduler/GenericScheduler.php:57]: > Do we want to create a lockfile, handle that in the DB or through cache?

That's the difficult question 😅 Database is out, since this package should work standalone without a DB. So either cache or a lock file. I think a lock file is better since we don't want the schedule to be lost on cache clear.

For now, I'd use a simple file like we do with discovery caching, we can always expand on it later and make it configurable, but that's out of scope for this PR

Comment (przemyslaw-przylucki) [reply src/Scheduler/GenericScheduler.php:57]: Makes sense! Will adjust it later today

File: src/Scheduler/CronDefinition.php
Comment (przemyslaw-przylucki) [reply src/Scheduler/CronDefinition.php]: Didn't know about Sagas, thanks for a nice read!

Why does it have a hacky feeling?

Well, what I wanted to do is create this syntax:

```php
#[Interval(minute: 1)]
public function prune(): void
{
   DB::table('messages')->where('created_at', '<=', now())->delete();
}
```

instead of 

```php
#[ConsoleCommand(name: 'abc', schedule: new DateTimeInterval('1Y'))
public function prune(): void
{
   DB::table('messages')->where('created_at', '<=', now())->delete();
}
```

which means that we'd have to be able to call a command from shell to execute that specific handler, but that wouldn't necessarily be a command, just something that runs every X time

File: src/Scheduler/GenericScheduler.php
Comment (przemyslaw-przylucki) [reply src/Scheduler/GenericScheduler.php]: Hey @brendt, reworked this a bit, still lacks tests and it doesn't handle sub-minute runs yet - is that something that we'd want to support from the get go or can we add it later?

Lemme know your thoughts!

File: src/Discovery/ScheduleDisovery.php
Comment (przemyslaw-przylucki) [inline src/Discovery/ScheduleDisovery.php:23]: Creating a separate discovery for it felt cleaner in some ways but also weird in others - not really sure how to feel about it

File: src/Scheduler/GenericScheduler.php
Comment (przemyslaw-przylucki) [inline src/Scheduler/GenericScheduler.php:45]: Do we want to acknowledge Windows existence at this point?

File: src/Scheduler/GenericScheduler.php
Comment (brendt) [reply src/Scheduler/GenericScheduler.php]: Nah we don't need sub-minute runs atm 👍

File: src/Inititalizers/SchedulerInitializer.php
Comment (brendt) [inline src/Inititalizers/SchedulerInitializer.php:32]: Following Tempest's style, I'd expect the `SchedulerConfig` to be injected here, and not a resolver

File: src/Discovery/ScheduleDisovery.php
Comment (brendt) [reply src/Discovery/ScheduleDisovery.php:23]: I'm fine with it, as long as it's tied to a separate config object, it's ok to be separate

File: src/Scheduler/GenericScheduler.php
Comment (brendt) [reply src/Scheduler/GenericScheduler.php:45]: Nope 👍

File: src/Scheduler/SchedulerConfig.php
Comment (brendt) [inline src/Scheduler/SchedulerConfig.php:29]: Why do we need to set the handler here? Isn't that done while building the console config object?

File: src/Scheduler/StatefulScheduledCommandsResolver.php
Comment (brendt) [inline src/Scheduler/StatefulScheduledCommandsResolver.php:43]: Could these methods be private?

File: src/Scheduler/StatefulScheduledCommandsResolver.php
Comment (brendt) [inline src/Scheduler/StatefulScheduledCommandsResolver.php:21]: `get` should probably be renamed to `getPreviousRuns()`. I'd also rename `lastRuns` to `previousRuns`

File: src/Scheduler/StatefulScheduledCommandsResolver.php
Comment (brendt) [inline src/Scheduler/StatefulScheduledCommandsResolver.php:19]: This should be a better name. I actually think all of this code should be moved to within `Scheduler`, until we decide to add multiple scheduling strategies — which is out of scope for now

File: src/Scheduler/StatefulScheduledCommandsResolver.php
Comment (brendt) [inline src/Scheduler/StatefulScheduledCommandsResolver.php:58]: Private?

File: src/Scheduler/ScheduledCommandsResolver.php
Comment (brendt) [inline src/Scheduler/ScheduledCommandsResolver.php:10]: As mentioned further down below, I don't think this interface should exist, and the code should be moved as private methods to `GenericScheduler`

File: src/Scheduler/SchedulerConfig.php
Comment (brendt) [inline src/Scheduler/SchedulerConfig.php:15]: Do `$output` and `$outputMode` have to be configurable?

File: src/Scheduler/GenericScheduler.php
Comment (brendt) [inline src/Scheduler/GenericScheduler.php:34]: I think I would prefer to have this scheduler not worry about writing output. Instead it should dispatch events that `schedule:run` hooks into, to write to the output. 

There's one big caveat though: events aren't included in `tempest/core` yet 😅 So feel free to ignore this comment, I'll fix it after merging this PR :)

File: src/Scheduler/GenericScheduler.php
Comment (brendt) [inline src/Scheduler/GenericScheduler.php:40]: I don't think it's finished at this point, since the command is run in the background. Running them as background jobs actually opens a huge can of worms, because now we need to keep track of running, succeeded, and failed commands as well.

That in turn leads to this feature becoming a queue manager instead of a simple scheduler. I think that's out of scope. (I do plan on writing a queue manager, but not within this package)

I think we should always run scheduled commands in the same process. But I might overlook some problems with that approach. I'd appreciate your input

**Review by brendt — CHANGES_REQUESTED** [review]

File: src/Scheduler/CronDefinition.php
Comment (brendt) [reply src/Scheduler/CronDefinition.php]: What I gathered from the PR, the syntax now looks like this

```php
#[Schedule(new DateInterval('P1D'))]
```

I agree that this isn't ideal. Let's brainstorm some options


---

```php
#[Schedule(new DateInterval('P1D'))]
```

This is the most flexible, but also the most confusing syntax

---

```php
#[Schedule(days: 3, hours: 1, minutes: 2)] // At least 3 days, 1 hour and 2 minutes have to be passed
```

A lot more comprehensible, we could even have an option here to run a command on a fixed time:

```php
#[Schedule(at: '10:20')]
```

---

```php
#[Schedule(cron: '3-59/15 6-12 */15 1 2-5')]
```

There are packages that parse cron expressions: https://github.com/dragonmantank/cron-expression

However, while super flexible, I don't think it's the most straight forward.

---

```php
#[Schedule(Every::minute(2))]
```

Not sure about it, just trying to come up with some other approaches.

File: src/Scheduler/GenericScheduler.php
Comment (przemyslaw-przylucki) [reply src/Scheduler/GenericScheduler.php:34]: Yeah for sure - this entire PR was more so to get directions, not merge it yet :D

File: src/Scheduler/GenericScheduler.php
Comment (przemyslaw-przylucki) [reply src/Scheduler/GenericScheduler.php:40]: It should say "dispatched" instead, but we don't really want this output here either way.

> I don't think it's finished at this point, since the command is run in the background. Running them as background jobs actually opens a huge can of worms, because now we need to keep track of running, succeeded, and failed commands as well.

I may be missing something here, but from the point of scheduler, we are not really interested in whether or not they succeed - we just want stuff to run, or really - be dispatched at specific time, if we were to execute them in sync, we'd have to deal with delays and stuff not running on time.

I think having them run in background should be the default option, defined in config.

I think that eventually, what would be cool is if we could make the Scheduler dispatch n-tasks to execute commands in parallel that'd report their status via events - this way we could have it work kinda both ways

File: src/Scheduler/CronDefinition.php
Comment (przemyslaw-przylucki) [reply src/Scheduler/CronDefinition.php]: Okay, I just want to clear a couple of things, so we have a shared understanding of the intended functionality.

> which means that we'd have to be able to call a command from shell to execute that specific handler, but that wouldn't necessarily be a command, just something that runs every X time

What I proposed is being able to add that attribute over *any* function in Tempest project, not just commands.

If we want this to work for any function, how would you approach handling that?

One approach would be to serialize the function and use the hash as a name to call from the scheduler, the only downside is that modifications would make it

If we want to make this work **only for commands**, is there a need to add a separate attribute for it, or could this just be another property on `ConsoleCommand`?

File: src/Scheduler/StatefulScheduledCommandsResolver.php
Comment (przemyslaw-przylucki) [reply src/Scheduler/StatefulScheduledCommandsResolver.php:43]: I'll refactor the whole thing later on - the intend of this was to listen to feedback on how you'd want this to work high-level first 😎

File: src/Scheduler/SchedulerConfig.php
Comment (przemyslaw-przylucki) [reply src/Scheduler/SchedulerConfig.php:31]: The way it works now is that discovery of schedules is independent from the ConsoleConfig, this will be more useful if we decide to go the "attribute over any function" path

File: src/Scheduler/ScheduledCommandsResolver.php
Comment (przemyslaw-przylucki) [reply src/Scheduler/ScheduledCommandsResolver.php:10]: We can def drop it if you want to - I've added it because you've mentioned that we should probably provide the user with option to do stateful or stateless scheduler dispatches - with this interface we won't have to change the `GenericScheduler` regardless of how the user may want to resolve commands to run, which seems like a nice thing to have tbh.

If we just want to go stateless in this PR without anticipating stateful to be implemented relatively soon, or at all, then it makes complete sense to drop it.

Lemme know your thoughts!

File: src/Scheduler/SchedulerConfig.php
Comment (przemyslaw-przylucki) [reply src/Scheduler/SchedulerConfig.php:16]: I think that they do have to be configurable - if the command's execution fails, we'd want to see what went wrong and be able to debug whether or not they executed at all, it's a configuration setting in every other project as far as I'm aware, I do believe that it should live at the #[Schedule] level though, where it was originally, but after your mixed feelings about it, this felt like a nice compromise :D 

Lemme know how you want to move forward with it

File: src/Scheduler/ScheduledCommandsResolver.php
Comment (przemyslaw-przylucki) [reply src/Scheduler/ScheduledCommandsResolver.php:10]: Okay I missed your other comment - will remove

File: src/Scheduler/SchedulerConfig.php
Comment (brendt) [reply src/Scheduler/SchedulerConfig.php:30]: Gotcha. Maybe then it makes more sense to keep everything within one discovery, not sure.

Let's keep it as is for now, we can always refactor later :)

File: src/Scheduler/SchedulerConfig.php
Comment (brendt) [reply src/Scheduler/SchedulerConfig.php:15]: Well now that you explained it like this, I agree that it makes most sense to have it on the `Schedule` attribute. So sorry for changing my mind 😅 😁

File: src/Scheduler/GenericScheduler.php
Comment (brendt) [reply src/Scheduler/GenericScheduler.php]: > if we were to execute them in sync, we'd have to deal with delays and stuff not running on time.

That's a good point 🤔

I'm just a little scared about things being dispatched to the background, and not having any way of interacting with them anymore. What if a process hangs indefinitely? We definitely need some good logging then. 

Actually. Writing this down. It would be pretty powerful if the scheduler and queue manager shared the same infrastructure. I always found it a bit confusing with Laravel that there is a basic scheduler, and then there's horizon as something separate.

I'm not saying we should write that queue management part right now, but it might be an interesting area to explore afterwards.

Ok, running in the background makes most sense. Thanks for sharing your thought!

**Comment by brendt** [comment]
> 1. Do we want any function to be schedule-able, or just commands?

Wow, interesting though. Very tempest-y as well. I suppose we'd have to determine whether a handler is a simple function, or whether it's a console command, and run it differently depending on that. 

Wow, that's a great idea. Yes, let's do this!

> 2. Where and do we keep the configuration for:

Well, with "scheduled" items becoming a separate thing, I think keeping everything within the `Schedule` attribute (with defaults) makes sense. This also answers the separate discovery question I had before.

> 3. How do we define intervals?

Yeah, let's go with the `Every` approach 👍 I'm not sure whether `Every::hour()->at(17)` is possible within attributes, we should play around with it.

I hope that, one day, we get [tagged unions](https://wiki.php.net/rfc/tagged_unions), which would mean we could attach this data to enums, but that shouldn't stop us for now :)

**Comment by brendt** [comment]
I think `#[Schedule(years: 1, minutes: 50)]` is the second best thing.

**Comment by brendt** [comment]
Or… what if we did it like this?

```
#[Schedule(Every::MINUTE)]
#[Schedule(Every::HOUR)]
#[Schedule(Every::DAY)]
#[Schedule(Every::WEEK)]
#[Schedule(Every::MONTH)]
````

This would be the default usage with the enum, I figure they cover the 90% use case. Next we allow more fine grained control:

```
#[Schedule(new Interval(days: 2, hours: 3))]
```

So schedule would look like this:

```php
#[Attribute]
class Schedule
{

     public Interval $interval;

    public function __construct(Every|Interval $interval)
    {
        $this->interval = $interval instanceof $interval ? $interval : $interval->getInterval()
    }
}
```

**Comment by przemyslaw-przylucki** [comment]
That's even nicer! I was playing around with this but was hesitant to suggest since https://stitcher.io/blog/opinion-driven-design :P 

Alright, I think I have everything I need for now - I'll try to tie it up in the next day or two and get back to you!

Thanks again for the feedback!

**Comment by brendt** [comment]
Ping me when you're ready :)

**Comment by brendt** [comment]
FYI, I already wrote a draft doc for this feature: https://tempest.stitcher.io/console/05-scheduler

Let me know if I should change any of the details.

**Comment by przemyslaw-przylucki** [comment]
@brendt feel free to tear this to pieces :D

I added a new abstraction for CommandExecutor, mostly for testing purposes, but it could be useful if people want to acknowledge Windows existence or do something else idk

When it comes to naming conventions - should enums be all uppercase?

**Comment by brendt** [comment]
Will review later today! 

I had a weird idea last night while falling asleep, I want to pitch it here, but we don't need to change this PR. Would love your input.

It's about writing output from within a background task. We want the output of any task to written to some kind of (configurable) log. But, ideally, we want this to happen transparently. What if… we create a new `Console` implementation that will write to a log file instead of STDOUT? This implementation would be swapped into the container when running in a background task. That way, anything could use `$console->writeln()` just as normal, and depending on whether it's run as a console command or in the background, the output will end up somewhere else.

We could also change the interactive methods within this implementation to write errors and throw exceptions, since we can't give user input in background tasks. 

Maybe this means we don't need configurable output anymore on each schedule attribute, instead we can make this console implementation configurable. I don't know if that's flexible enough.

File: src/Scheduler/GenericInvocationExecutor.php
Comment (brendt) [inline src/Scheduler/GenericInvocationExecutor.php:7]: I think I want to rework the whole invocation structure. Instead of going through the shell again, we can build the command, and then pass it to the `ConsoleApplication` without having to boot tempest once more. 

This is something I'll refactor after merge though, as it might involve some bigger changes.

Comment (brendt) [reply src/Scheduler/GenericInvocationExecutor.php:7]: Edit: maybe it won't be possible, because we want to run stuff in the background. Ok, I'll just play around with it and see where it leads me

**Review by brendt — APPROVED** [review]
Left some comments, but those are just for me to review after merge. 

I plan on merging this later today, and then play around with it and finetune it some more.

Great work 💯

**Comment by brendt** [comment]
🥳

**Comment by przemyslaw-przylucki** [comment]
Let's goooo! 🎉 

Few things to consider to expand on in the future:
- adding sub-minute scheduler (should be very simple, I even pushed it here earlier but decided to drop it to have the initial PR simpler)
- adding mutexes (after cache/redis layer is done) so commands don't overlap

Will address the rest of the comments later today, I think that there's still some interesting plays we could do here :D

**Pull request merged**
